daemon/stop.go

```diff
@@ -38,52 +38,56 @@ func (daemon *Daemon) ContainerStop(name string, timeout *int) error {
 
 // containerStop sends a stop signal, waits, sends a kill signal.
 func (daemon *Daemon) containerStop(container *containerpkg.Container, seconds int) error {
+	// TODO propogate a context down to this function
+	ctx := context.TODO()
 	if !container.IsRunning() {
 		return nil
 	}
-
+	var wait time.Duration
+	if seconds >= 0 {
+		wait = time.Duration(seconds) * time.Second
+	}
+	success := func() error {
+		daemon.LogContainerEvent(container, "stop")
+		return nil
+	}
 	stopSignal := container.StopSignal()
-	// 1. Send a stop signal
-	if err := daemon.killPossiblyDeadProcess(container, stopSignal); err != nil {
-		// While normally we might "return err" here we're not going to
-		// because if we can't stop the container by this point then
-		// it's probably because it's already stopped. Meaning, between
-		// the time of the IsRunning() call above and now it stopped.
-		// Also, since the err return will be environment specific we can't
-		// look for any particular (common) error that would indicate
-		// that the process is already dead vs something else going wrong.
-		// So, instead we'll give it up to 2 more seconds to complete and if
-		// by that time the container is still running, then the error
-		// we got is probably valid and so we force kill it.
-		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
-		defer cancel()
 
-		if status := <-container.Wait(ctx, containerpkg.WaitConditionNotRunning); status.Err() != nil {
-			logrus.Infof("Container failed to stop after sending signal %d to the process, force killing", stopSignal)
-			if err := daemon.killPossiblyDeadProcess(container, 9); err != nil {
-				return err
-			}
-		}
+	// 1. Send a stop signal
+	err := daemon.killPossiblyDeadProcess(container, stopSignal)
+	if err != nil {
+		wait = 2 * time.Second
 	}
 
-	// 2. Wait for the process to exit on its own
-	ctx := context.Background()
+	subCtx, cancel := context.WithCancel(ctx)
 	if seconds >= 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, time.Duration(seconds)*time.Second)
-		defer cancel()
+		// if seconds has been set, then set a timeout on the context so we dont wait forever
+		subCtx, cancel = context.WithTimeout(ctx, wait)
+	}
```

Comment:
If `seconds` is 0, even if we set a value for `wait` on line 59 we're still not setting a timeout on the context?  Is that intentional?  Would it be better to check whether `wait` is zero here?

Reply:
`docker stop` explicitly allows disabling the `docker kill` fallback by setting a negative timeout value (and thus waiting forever after SIGTERM). That's why it's checking `>=`.

If the user explicitly set the timeout for the wait after sending a stop signal to 0, then I guess we should respect that.

Reply:
> If the user explicitly set the timeout for the wait after sending a stop signal to 0, then I guess we should respect that.

You're doing that on line 59 irrespective of whether `seconds` was `>= 0` which is why I'm confused.

Reply:
see comment below: https://github.com/moby/moby/pull/41586#issuecomment-786871667

Reply:
yep that's correct